### PR TITLE
[Widget Inspector] Fix stack overflow error for Flutter web when requesting a large widget tree

### LINKED ISSUE #8553: [Inspector 2.0] Stack overflow error when building the new widget tree with implementation widgets
A Flutter web customer with a large widget tree reported that the new inspector is crashing for them when "show implementation widgets" is toggled on. 

When it crashes, they see a "stack overflow" error in the Chrome DevTools console. I've confirmed that this is due to us building the widget tree's JSON object recursively, and it seems like we are hitting Chrome's stack limit. 

They've confirmed that they don't see the stack overflow error when the widget tree JSON is built iteratively instead.

### PR DESCRIPTION
**Pull request by elliette**
Fixes https://github.com/flutter/devtools/issues/8553

Context: 

A Flutter web customer with a large widget tree was getting a stack overflow error when they toggled on "show implementation widgets" in the Flutter DevTools Inspector. This is because building the JSON tree recursively was hitting Chrome's stack limit. 

This PR creates the JSON tree **iteratively** if the `getRootWidgetTree` service extension is called with `fullDetails = false` (which is what DevTools uses to fetch the widget tree).

For all other instances of creating a widget JSON map (for example, when fetching widget properties) the recursive implementation is used. This allows properties provided by subclasses implementing `toJsonMap` to be included in the response.

Note: Because with this change `toJsonMap` is only called when `fullDetails = true` and `toJsonMapIterative` is only called when `fullDetails = false`, this PR partially reverts the changes in https://github.com/flutter/flutter/pull/157309.

### CODE REVIEW

File: packages/flutter/lib/src/foundation/diagnostics.dart [hunk 1]
```diff
@@ -1605,29 +1616,12 @@ abstract class DiagnosticsNode {
...
-    bool fullDetails = true,
-  }) {
+  Map<String, Object?> toJsonMap(DiagnosticsSerializationDelegate delegate) {
     Map<String, Object?> result = <String, Object?>{};
     assert(() {
       final bool hasChildren = getChildren().isNotEmpty;
-      final Map<String, Object?> essentialDetails = <String, Object?>{
+      result = <String, Object?>{
         'description': toDescription(),
-        'shouldIndent': style != DiagnosticsTreeStyle.flat &&
```
Comment (kenzieschmoll) [inline packages/flutter/lib/src/foundation/diagnostics.dart:1617]: do we need to add this back to the result?

File: packages/flutter/lib/src/foundation/diagnostics.dart [hunk 2]
```diff
@@ -1651,19 +1645,53 @@ abstract class DiagnosticsNode {
...
+  /// configuration provided in the [DiagnosticsSerializationDelegate].
+  ///
+  /// This is only used when [WidgetInspectorServiceExtensions.getRootWidgetTree]
+  /// is called with fullDetails=false. To get the full widget details, including
+  /// any details provided by subclasses, [toJsonMap] should be used instead.
+  ///
+  /// See https://github.com/flutter/devtools/issues/8553 for details about this
+  /// iterative approach.
+  Map<String, Object?> toJsonMapIterative(
+      DiagnosticsSerializationDelegate delegate) {
```
Comment (kenzieschmoll) [inline packages/flutter/lib/src/foundation/diagnostics.dart:1677]: trailing comma?

File: packages/flutter/lib/src/foundation/diagnostics.dart [hunk 3]
```diff
@@ -1803,6 +1827,73 @@ abstract class DiagnosticsNode {
...
+  Map<String, Object?> _toJson(
+    DiagnosticsSerializationDelegate delegate, {
+    required _NodesToJsonifyStack childrenToJsonify,
+  }) {
+    final List<_JsonDiagnosticsNode> childrenJsonList =
+        <_JsonDiagnosticsNode>[];
+    final bool includeChildren =
+        getChildren().isNotEmpty && delegate.subtreeDepth > 0;
+
+    // Collect the children nodes to convert to JSON later:
```
Comment (kenzieschmoll) [inline packages/flutter/lib/src/foundation/diagnostics.dart:1857]: nit: remove colon at end of comment

File: packages/flutter/lib/src/foundation/diagnostics.dart [hunk 4]
```diff
@@ -1803,6 +1827,73 @@ abstract class DiagnosticsNode {
...
+
+  void _jsonifyNextNodesInStack(
+    _NodesToJsonifyStack toJsonify, {
+    required DiagnosticsSerializationDelegate delegate,
+  }) {
+    while (toJsonify.isNotEmpty) {
+      final (
+        DiagnosticsNode nextNode,
+        void Function(_JsonDiagnosticsNode) callback
+      ) = toJsonify.removeAt(0);
```
Comment (kenzieschmoll) [inline packages/flutter/lib/src/foundation/diagnostics.dart:1839]: if we are always removing from the front, should we use a more efficient data structure than a list like a ListQueue?

File: packages/flutter/lib/src/foundation/diagnostics.dart (hunk 1 above)
Comment (kenzieschmoll) [reply packages/flutter/lib/src/foundation/diagnostics.dart:1617]: nevermind I see it is added below.

File: packages/flutter/lib/src/foundation/diagnostics.dart (hunk 3 above)
Comment (elliette) [reply packages/flutter/lib/src/foundation/diagnostics.dart]: Removed (changed to period)

File: packages/flutter/lib/src/foundation/diagnostics.dart (hunk 4 above)
Comment (elliette) [reply packages/flutter/lib/src/foundation/diagnostics.dart]: Ooh I didn't know ListQueue exists, that's perfect

**Review by kenzieschmoll — APPROVED** [review]

**Pull request merged**